### ISSUE TITLE
Lifetime bounds on structs do not entirely constrain impl fns

### ISSUE THREAD

**Issue by talchas** [rust-lang/rust#28609]
```rust
struct Foo<'a, 'b: 'a>(&'a &'b i32);
impl<'a, 'b> Foo<'a, 'b> {
    fn foo(&self) {}
}
```
compiles. As far as I can tell, the bound on 'b is checked in the definitions of the functions, but isn't checked in the use of them. As a concrete example of this Kimundi/scoped-threadpool-rs#8 ( https://play.rust-lang.org/?gist=870da1405ea15e80f778&version=nightly ) will allow the incorrect use of ScopeRef in the test, while copying the bounds from the struct definition to the impl definition correctly rejects it.

**Comment by bluss** [comment]
cc @nikomatsakis

**Comment by arielb1** [comment]
Minified:
```Rust
struct S<'a, 'b: 'a>(Option<&'a &'b ()>);

impl<'a, 'b> S<'a, 'b> {
    fn xform(&self, a: &'b mut u32) -> &'a mut u32 {
        a
    }
}

fn return_dangling_pointer<'a>(s: S<'a, 'a>) -> &'a mut u32 {
    s.xform(&mut 3)
}

fn main() {
    let s = S(None);
    let a = return_dangling_pointer(s);
    println!("{}", a);
}
```

**Comment by arielb1** [comment]
If you use a trait impl or UFCS instead of an inherent impl, the error is reported correctly.

**Comment by arielb1** [comment]
This is a nasty soundness bug. I think its a just-a-bug I can fix.

**Comment by arielb1** [comment]
Can also be done via `Deref`

```Rust
use std::ops::Deref;

struct S<'a, 'b: 'a>(Option<&'a &'b ()>, &'b u32);

impl<'a, 'b> Deref for S<'a, 'b> {
    type Target = &'a u32;
    fn deref(&self) -> &&'a u32 {
        &self.1
    }
}

fn return_dangling_pointer<'a>(s: S<'a, 'a>) -> &'a u32 {
    let four = 4;
    let mut s = s;
    s.1 = &four;
    &s // or &*s
}

fn main() {
    let temp = &42;
    let ptr = return_dangling_pointer(S(None,&temp));
    println!("{}", ptr);
}
```

**Comment by arielb1** [comment]
Also via overloaded ops:

```Rust
use std::ops::Shl;

struct S<'a, 'b: 'a>(Option<&'a &'b ()>);

impl<'a, 'b> Shl<&'b u32> for S<'a, 'b> {
    type Output = &'a u32;
    fn shl(self, t: &'b u32) -> &'a u32 { t }
}

fn return_dangling_pointer<'a>(s: S<'a, 'a>) -> &'a u32 {
    let s = s;
    s << &mut 3 /* avoid promotion */
}

fn main() {
    let a = return_dangling_pointer(S(None));
    println!("{}", a);
}
```

**Comment by nikomatsakis** [comment]
triage: P-high
